torch/_dynamo/variables/inline_helper.py

```diff
@@ -0,0 +1,167 @@
+import warnings
+
+import torch.fx
+import torch.fx.traceback as fx_traceback
+import torch.utils._pytree as pytree
+from torch.fx import Interpreter
+from torch.nn.utils import stateless
+from .constant import ConstantVariable
+
+
+def dummy_user_function_to_inline_gm(gm, args):
+    return gm(*args)
+
+
+def dummy_user_function_to_inline_wrapped_gm(wrapped_gm, args, kwargs):
+    return wrapped_gm(args, kwargs)
+
+
+def dummy_accumulate_grad_(t1, t2):
+    if t1.grad is None:
+        t1.grad = t2
+    else:
+        t1.grad += t2
+
+
+# copy from torch/_functorch/_aot_autograd/traced_function_transforms.py
+# and remove the restriction that output has to be a tuple.
+def create_functional_call(mod, params_spec, params_len, store_orig_mod=False):
+    # Redundant with dynamo, but worth having in case this gets invoked elsewhere.
+    # https://github.com/pytorch/pytorch/issues/103569
+
+    def functional_call(*args, **kwargs):
+        with stateless._reparametrize_module(
+            mod, pytree.tree_unflatten(args[:params_len], params_spec)
+        ):
+            if isinstance(mod, torch.fx.GraphModule):
+                with fx_traceback.preserve_node_meta(), warnings.catch_warnings():
+                    warnings.filterwarnings(
+                        "ignore", "Anomaly Detection has been enabled."
+                    )
+                    with torch.autograd.detect_anomaly(check_nan=False):
+                        out = Interpreter(mod).run(*args[params_len:], **kwargs)
+            else:
+                out = mod(*args[params_len:], **kwargs)
+        return out
+
+    # Note [Preserving the nn module stack metadata during export non-strict mode]
+    # This path is currently only used by the non-strict export flow,
+    # where we cannot rely on dynamo to preserve nn stack metadata in our captured graph.
+    # Instead, we stash the original user nn module here, and rely on `make_fx` to grab
+    # this stashed module and use it to track nn module stack metadata
+    if store_orig_mod and not hasattr(functional_call, "_orig_mod"):
+        functional_call._orig_mod = mod  # type: ignore[attr-defined]
+
+    return functional_call
+
+
+code_to_Fx = {}
+
+
+def vt_to_fake_helper(vt, tx):
+    from ..utils import get_fake_value
+
+    proxy_ = vt.as_proxy()
+
+    def proxy_to_fake_helper(p):
+        if type(p) is torch.fx.proxy.Proxy:
+            return get_fake_value(p.node, tx)
+        elif type(p) is tuple:
+            return tuple(map(proxy_to_fake_helper, p))
+        else:
+            # mostly handle scalar
+            # check return type is a fake tensor
+            assert type(p) != torch.fx.proxy.Proxy
+            return p
+
+    return proxy_to_fake_helper(proxy_)
+
+
+def decompose_and_inline_function_with_makefx(tx, fn, args, kwargs, function_key=None):
```

Comment:
@anijain2305 The other thing is that this PR is both inlining into NN modules, and also taking the IR all the way to pre-dispatch ATen IR. I suppose you could make the argument that going only half way is potentially useful, but @JackCaoG's workstream needs to keep going, all the way to inlining past composite implicit autograd, so all tensors which will be visible for autograd are visible. So he will still end up having to do this eventually.

Reply:
Oh I see you discussed this above in comments